src/gsy_e/models/strategy/energy_parameters/heatpump/cop_models/cop_models.py

```diff
@@ -0,0 +1,141 @@
+from typing import Optional
+from logging import getLogger
+
+from gsy_framework.enums import HeatPumpSourceType
+
+log = getLogger(__name__)
+
+
+class COPModelType(Enum):
+    """Selection of supported COP models"""
+
+    UNIVERSAL = 0
+    ELCO_AEROTOP_S09_IR = 1
+    ELCO_AEROTOP_G07_14M = 2
+    HOVAL_ULTRASOURCE_B_COMFORT_C11 = 3
+
+
+MODEL_FILE_DIR = os.path.join(os.path.dirname(__file__), "model_data")
+
+MODEL_TYPE_FILENAME_MAPPING = {
+    COPModelType.ELCO_AEROTOP_S09_IR: "Elco_Aerotop_S09M-IR_model_parameters.json",
+    COPModelType.ELCO_AEROTOP_G07_14M: "Elco_Aerotop_G07-14M_model_parameters.json",
+    COPModelType.HOVAL_ULTRASOURCE_B_COMFORT_C11: "hoval_UltraSource_B_comfort_C_11_model_"
+    "parameters.json",
+}
+
+
+class BaseCOPModel:
+    """Base clas for COP models"""
+
+    @abstractmethod
+    def calc_cop(self, source_temp_C: float, tank_temp_C: float, heat_demand_kW: Optional[float]):
+        """Return COP value for provided inputs"""
+
+
+class IndividualCOPModel(BaseCOPModel):
+    """Handles cop models for specific heat pump models"""
+
+    def __init__(self, model_type: COPModelType):
+        with open(
+            os.path.join(MODEL_FILE_DIR, MODEL_TYPE_FILENAME_MAPPING[model_type]),
+            "r",
+            encoding="utf-8",
+        ) as fp:
+            self._model = json.load(fp)
+        self.model_type = model_type
+
+    def _calc_power(self, source_temp_C: float, tank_temp_C: float, heat_demand_kW: float):
+        CAPFT = (
+            self._model["CAPFT"][0]
+            + self._model["CAPFT"][1] * source_temp_C
+            + self._model["CAPFT"][3] * source_temp_C**2
+            + self._model["CAPFT"][2] * tank_temp_C
+            + self._model["CAPFT"][5] * tank_temp_C**2
+            + self._model["CAPFT"][4] * source_temp_C * tank_temp_C
+        )
+
+        HEIRFT = (
+            self._model["HEIRFT"][0]
+            + self._model["HEIRFT"][1] * source_temp_C
+            + self._model["HEIRFT"][3] * source_temp_C**2
+            + self._model["HEIRFT"][2] * tank_temp_C
+            + self._model["HEIRFT"][5] * tank_temp_C**2
+            + self._model["HEIRFT"][4] * source_temp_C * tank_temp_C
+        )
+
+        # Partial Load Ratio (PLR)
+        PLR = heat_demand_kW / (self._model["Qref"] * CAPFT)
+
+        # HEIRFPLR calculation
+        HEIRFPLR = (
+            self._model["HEIRFPLR"][0]
+            + self._model["HEIRFPLR"][1] * PLR
+            + self._model["HEIRFPLR"][2] * PLR**2
+        )
+
+        # Power consumption (P) calculation
+        return self._model["Pref"] * CAPFT * HEIRFT * HEIRFPLR
+
+    def calc_cop(self, source_temp_C: float, tank_temp_C: float, heat_demand_kW: float):
+        assert heat_demand_kW is not None, "heat demand should be provided"
+        if heat_demand_kW == 0:
+            return 0
+        electrical_power_kW = self._calc_power(source_temp_C, tank_temp_C, heat_demand_kW)
+        if electrical_power_kW <= 0:
+            log.debug(
```

Comment:
I would write an error here, and also I would prepend the configuration / simulation id in case this error will be logged in deployed environment.

Reply:
This might clutter or logs, but OK, let's do it and see how often this happens.